Sources/NIOCore/AsyncSequences/NIOBackPressuredAsyncSequence.swift

```diff
@@ -0,0 +1,1003 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIOConcurrencyHelpers
+
+#if compiler(>=5.5.2) && canImport(_Concurrency)
+/// A protocol for the back-pressure strategy of the ``NIOBackPressuredAsyncSequence``.
+///
+/// A back-pressure strategy is invoked when new elements are yielded to the sequence or
+/// when an ``NIOBackPressuredAsyncSequence/AsyncIterator`` requested the next value. The responsibility of the strategy is
+/// to determine wether more elements need to be demanded .
+///
+/// If more elements need to be requested, either the ``NIOBackPressuredAsyncSequenceSourceDelegate/demand()``
+/// method will be called or a ``NIOBackPressuredAsyncSequence/Source/YieldResult`` will be returned that indicates
+/// to demand more.
+///
+/// - Important: The methods of this protocol are guaranteed to be called serially. Furthermore, the implementation of these
+/// methods should not do any locking or call out to any other Task/Thread.
+public protocol NIOBackPressuredAsyncSequenceStrategy {
+    /// This method is called after new elements were yielded to the source.
+    ///
+    /// - Parameter bufferDepth: The current depth of the internal buffer.
+    /// - Returns: Returns wether more elements should be demanded.
+    mutating func didYield(bufferDepth: Int) -> Bool
+
+    /// This method is called after `next` was called on an iterator of the ``NIOBackPressuredAsyncSequence``.
+    ///
+    /// - Parameter bufferDepth: The current depth of the internal buffer.
+    /// - Returns: Returns wether new elements should be demanded.
+    mutating func didNext(bufferDepth: Int) -> Bool
+}
+
+/// The delegate of ``NIOBackPressuredAsyncSequence``.
+///
+/// - Important: All methods of this protocol are called inside the `Lock` of the ``NIOBackPressuredAsyncSequence``.
```

Comment:
OOF that is a steep requirement... is there a way we can make these isolated instead? similar to how we implemented AsyncBuffer in the async algorithms package?

Reply:
The `AsyncBuffer` is an `actor` which we can't really use here because that would require `async` calls to the strategy.
In the end, we need something is running inside the state machine of the `NIOBackPressuredAsyncSequence` since it is basically an extension of our internal state but pluggable. 

I agree with you that this can be quite dangerous to hold, that's why we intend to supply two blessed implementations. We don't expect people to generally write their own strategies. cc @Lukasa